### ISSUE TITLE
Terraform hangs when creating a group

### ISSUE THREAD

**Issue by max-rocket-internet** [mrparkers/terraform-provider-keycloak#487]
My TF code looks like this:

```hcl
provider "keycloak" {
    client_id     = "terraform"
    realm         = "aws"
    url           = "https://<keycloak host>"
}
resource "keycloak_group" "group" {
  realm_id = "aws"
  name     = "max-test"
}
```

`terraform init` completes fine but when I try to run `terraform apply` it just hangs. If I set env var `TF_LOG=TRACE` then I see this:

```
2021-02-18T16:18:10.904+0100 [DEBUG] plugin.terraform-provider-keycloak_v2.2.0: 2021/02/18 16:18:10 [DEBUG] Login response: {"access_token":"xxxx", "token_type":"bearer","not-before-policy":0,"session_state":"xxxxx","scope":"email profile"}
2021-02-18T16:18:10.904+0100 [DEBUG] plugin.terraform-provider-keycloak_v2.2.0: 2021/02/18 16:18:10 [DEBUG] Sending GET to /auth/admin/serverinfo
2021-02-18T16:18:10.967+0100 [DEBUG] plugin.terraform-provider-keycloak_v2.2.0: 2021/02/18 16:18:10 [DEBUG] Response: 403 Forbidden.  Attempting refresh
2021/02/18 16:18:15 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/18 16:18:15 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "keycloak_group.group (expand)"
2021/02/18 16:18:15 [TRACE] dag/walk: vertex "keycloak_group.group (expand)" is waiting for "provider[\"registry.terraform.io/mrparkers/keycloak\"]"
2021/02/18 16:18:15 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/mrparkers/keycloak\"] (close)" is waiting for "keycloak_group.group (expand)"
2021/02/18 16:18:20 [TRACE] dag/walk: vertex "keycloak_group.group (expand)" is waiting for "provider[\"registry.terraform.io/mrparkers/keycloak\"]"
2021/02/18 16:18:20 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "keycloak_group.group (expand)"
2021/02/18 16:18:20 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/mrparkers/keycloak\"] (close)" is waiting for "keycloak_group.group (expand)"
2021/02/18 16:18:20 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
```

Then those last few messages are just repeating. I don't really know much about keycloak to be honest but I created the client like the instructions, it looks like this:

![Screenshot 2021-02-18 at 16 25 00](https://user-images.githubusercontent.com/8859277/108379239-07227100-7206-11eb-89c8-e7393c54dc33.png)

**Comment by max-rocket-internet** [comment]
Any idea what I'm missing here?

**Comment by mrparkers** [comment]
Hey @max-rocket-internet, sorry for the late response.

The problem was related to a mutex that was used during the initial login and refresh process.  The initial login process locks a mutex, then fetches Keycloak server info via the `/serverinfo` endpoint.  This request seemed to fail for you, which invoked the refresh process, which utilizes that same mutex.  I removed the mutex usage altogether via #489, so process should no longer hang for you.

However, that initial failing request to `/serverinfo` will still cause you problems.  Make sure that the client you're logging in as has the ability to use this endpoint.  I've never seen this fail before, so I'm not 100% sure I know the cause, but we can work through it if its still giving you problems.

I'll be cutting a release soon, which you can try out to see if this issue is resolved.

**Comment by mrparkers** [comment]
I just pushed up a release for `v2.3.0` - do you want to give that a shot and see if that helps with your issue?

**Comment by max-rocket-internet** [comment]
@mrparkers I will test next week it and let you know.

> Make sure that the client you're logging in as has the ability to use this endpoint

How do I do that?

We switched to use this and now it works but I'd rather not use the admin user:

```
provider "keycloak" {
  client_id = "admin-cli"
  username  = "admin"
  url       = "https://xxxxx"
}
```

**Comment by mrparkers** [comment]
Honestly, I didn't even realize that the `/serverinfo` endpoint required some sort of role that wasn't given to a service account user by default.  So I'm not entirely sure 😕 .

To be clear, the `/serverinfo` endpoint is needed by the provider in order to know what features may or may not be supported by the Keycloak server it's working against.  I might be able to add another attribute to the provider like `server_version`, which could disable the `/serverinfo` check and use the version provided by the user.  Would that work for you?

**Comment by max-rocket-internet** [comment]
Hi @mrparkers 

OK great, with the new version I get an error quickly:

```
$ terraform plan

Error: error initializing keycloak provider

error sending GET request to /auth/admin/serverinfo: 403 Forbidden. Response
body: {"error":"unknown_error"}
```

>  I might be able to add another attribute to the provider like server_version, which could disable the /serverinfo check and use the version provided by the user. Would that work for you?

I don't think this is required, I think the client just needs the correct "roles". I did some testing, for example on the `Service Accounts Roles` tab, if you add under `Client Roles`, `realm-management` and role `view-identity-providers` then there's a different error. If I add `realm-admin` then it all works fine.

I think the main pain point for me is how utterly terrible the keycloak UI is (which was what led me to using TF and here in the first place) so it's not clear how to do step 3 from [here](https://github.com/mrparkers/terraform-provider-keycloak/blob/master/docs/index.md#client-credentials-grant-setup-recommended)

So I think what would be great is if we could update the instructions to be a bit clearer? How does this look https://github.com/mrparkers/terraform-provider-keycloak/pull/494 ?

**Comment by mrparkers** [comment]
I think this issue has been resolved with #489 and #494, so I'm going to close this.  Thanks again for the bug report!
